bindinfo/handle.go

```diff
@@ -0,0 +1,270 @@
+package bindinfo
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv/oracle"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/sqlexec"
+)
+
+// BindHandle is used to handle all the sql bind operations.
+type BindHandle struct {
+	sctx struct {
+		sync.Mutex
+		sessionctx.Context
+	}
+
+	// bindInfo caches the sql bind info from storage.
+	//
+	// The Mutex protects that there is only one goroutine changes the content
+	// of atmoic.Value.
+	//
+	// NOTE: Concurrent Value Write:
+	//
+	//    bindInfo.Lock()
+	//    newCache := bindInfo.Value.Load()
+	//    do the write operation on the newCache
+	//    bindInfo.Value.Store(newCache)
+	//
+	// NOTE: Concurrent Value Read:
+	//
+	//    cache := bindInfo.Load().
+	//    read the content
+	//
+	bindInfo struct {
+		sync.Mutex
+		atomic.Value
+	}
+
+	parser         *parser.Parser
+	lastUpdateTime types.Time
+}
+
+// NewBindHandle creates a new BindHandle.
+func NewBindHandle(ctx sessionctx.Context, parser *parser.Parser) *BindHandle {
+	handle := &BindHandle{parser: parser}
+	handle.sctx.Context = ctx
+	handle.bindInfo.Value.Store(make(cache, 32))
+	return handle
+}
+
+// Update updates the global sql bind cache.
+func (h *BindHandle) Update(fullLoad bool) (err error) {
+	sql := "select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info"
+	if !fullLoad {
+		sql += " where update_time > \"" + h.lastUpdateTime.String() + "\""
+	}
+
+	// No need to acquire the session context lock for ExecRestrictedSQL, it
+	// uses another background session.
+	rows, _, err := h.sctx.Context.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(nil, sql)
+	if err != nil {
+		return err
+	}
+
+	// Make sure there is only one goroutine writes the cache.
+	h.bindInfo.Lock()
+	newCache := h.bindInfo.Value.Load().(cache).copy()
+	defer func() {
+		if err == nil {
+			h.bindInfo.Value.Store(newCache)
+			h.bindInfo.Unlock()
+		}
+	}()
+
+	for _, row := range rows {
+		hash, meta, err := h.newBindMeta(newBindRecord(row))
+		if err != nil {
+			return err
+		}
+
+		newCache.removeStaleBindMetas(hash, meta)
+		newCache[hash] = append(newCache[hash], meta)
+
+		// Update lastUpdateTime to the newest one.
+		if meta.UpdateTime.Compare(h.lastUpdateTime) > 0 {
```

Comment:
If we successfully add several bind records, but the last fails, then we will not store the success records in cache, and we can never read the success records anymore because the update time is already updated.